From: Sergey Senozhatsky <senozhatsky@chromium.org>
Subject: zsmalloc: make sure we select best zspage size
Date: Tue, 25 Oct 2022 01:12:13 +0900

We currently decide the best zspage size by looking at used percentage
value.  This is not exactly enough as zspage usage percentage calculation
is not accurate.

For example, let's look at size class 208

pages per zspage       wasted bytes         used%
       1                   144               96
       2                    80               99
       3                    16               99
       4                   160               99

We will select 2 page per zspage configuration, as it is the first one to
reach 99%.  However, 3 pages per zspage wastes less memory.  Hence we need
to also consider wasted space metrics when device zspage size.

Additionally, rename max_usedpc_order because it does not hold zspage
order, it holds maximum pages per-zspage value.

Link: https://lkml.kernel.org/r/20221024161213.3221725-7-senozhatsky@chromium.org
Signed-off-by: Sergey Senozhatsky <senozhatsky@chromium.org>
Cc: Minchan Kim <minchan@kernel.org>
Cc: Nitin Gupta <ngupta@vflare.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/zsmalloc.c |   14 ++++++++++----
 1 file changed, 10 insertions(+), 4 deletions(-)

--- a/mm/zsmalloc.c~zsmalloc-make-sure-we-select-best-zspage-size
+++ a/mm/zsmalloc.c
@@ -775,8 +775,9 @@ out:
 static int get_pages_per_zspage(struct zs_pool *pool, int class_size)
 {
 	int i, max_usedpc = 0;
-	/* zspage order which gives maximum used size per KB */
-	int max_usedpc_order = 1;
+	/* zspage size which gives maximum used size per KB */
+	int pages_per_zspage = 1;
+	int min_waste = INT_MAX;
 
 	for (i = 1; i <= pool->max_pages_per_zspage; i++) {
 		int zspage_size;
@@ -788,14 +789,19 @@ static int get_pages_per_zspage(struct z
 
 		if (usedpc > max_usedpc) {
 			max_usedpc = usedpc;
-			max_usedpc_order = i;
+			pages_per_zspage = i;
 		}
 
 		if (usedpc == 100)
 			break;
+
+		if (waste < min_waste) {
+			min_waste = waste;
+			pages_per_zspage = i;
+		}
 	}
 
-	return max_usedpc_order;
+	return pages_per_zspage;
 }
 
 static struct zspage *get_zspage(struct page *page)
_
